### PR TITLE
feat(backend): direct injection 使用原始文档替代分块

### PR DESCRIPTION
**Pull request by sunnights**
## Summary by CodeRabbit

* **Bug Fixes**
  * Truncated or oversized source documents reliably trigger safe fallbacks to RAG to avoid incomplete direct injections; token/truncation checks tightened.

* **Refactor**
  * Routing now delegates direct-injection handling, centralizes RAG retrieval, applies metadata-driven immediate RAG, and emits clearer accept vs fallback telemetry.

* **Tests**
  * Expanded coverage for original-document retrieval, truncation rejection (boundary cases), routing decisions, and direct-injection vs RAG fallback scenarios.

### CODE REVIEW

File: backend/app/services/rag/retrieval_service.py [hunk 1]
```diff
@@ -240,6 +220,145 @@ def _get_direct_injection_rejection_reason(
...
+            direct_records
+        )
+        rejection_reason = self._get_direct_injection_rejection_reason(
+            route_mode=route_mode,
+            direct_records=direct_records,
+            direct_injection_estimated_tokens=direct_injection_estimated_tokens,
+            available_injection_tokens=available_injection_tokens,
+            context_window=context_window,
+            max_direct_chunks=max_direct_chunks,
+        )
```
[the remark this answers is not in the document]
Comment (sunnights) [reply backend/app/services/rag/retrieval_service.py:262]: 当前已有两层检查机制：                                                                                                                                                                                          
                                                                                                                                                                                                                  
  retrieve_with_routing
  ├── 第一层：_estimate_total_tokens_for_knowledge_bases (粗略估算)
  │   └── 使用 SUM(text_length) * 1.5，无需读取完整内容
  │   └── 如果超出 context_window → 直接跳过 direct injection
  │
  └── _try_direct_injection
      ├── get_original_documents_from_knowledge_base
      │   └── read_documents() 获取完整内容
      ├── _estimate_direct_injection_tokens (精确计算)
      └── _get_direct_injection_rejection_reason (精确检查)

  优化建议忽略的关键点：

  1. 粗略检查已经存在：_estimate_total_tokens_for_knowledge_bases 在进入 _try_direct_injection 之前就已经用元数据做了估算
  2. 精确检查需要完整内容：
    - max_direct_chunks 检查 → 可以用元数据（文档数量）
    - runtime_budget_exceeded 检查 → 需要精确 token 数
    - context_ratio_exceeded 检查 → 需要精确 token 数
  3. 边缘情况才需要精确检查：只有当粗略估算通过但实际超出时，才需要精确检查拒绝

File: backend/app/services/rag/retrieval_service.py [hunk 2]
```diff
@@ -714,6 +754,180 @@ async def _execute_runtime_query(
...
+
+            all_document_ids = [row[0] for row in doc_rows]
+            results = document_read_service.read_documents(
+                db=db,
+                document_ids=all_document_ids,
+                offset=0,
+                limit=10_000_000,
+                knowledge_base_ids=knowledge_base_ids,
+            )
+            records = self._build_document_records(results, knowledge_base_ids)
```
[the remark this answers is not in the document]
Comment (sunnights) [reply backend/app/services/rag/retrieval_service.py:846]: - 第一层检查（text_length >= MAX_EXTRACTED_TEXT_LENGTH）已足够
  - read_documents 调用时 limit=10_000_000，远大于正常文档
  - 第二层检查几乎不会触发，增加复杂度无实际收益
  - 不做防御性过度工程

File: backend/tests/services/rag/test_direct_injection.py [hunk 3]
```diff
@@ -0,0 +1,660 @@
...
+                query="test",
+                knowledge_base_ids=[123, 456],
+                db=db,
+                context_window=10000,
+                user_id=7,
+            )
+
+        # Should fallback to RAG because second KB has truncated documents
+        assert result["mode"] == "rag_retrieval"
+        service.retrieve_from_knowledge_base_internal.assert_awaited()
```
[the remark this answers is not in the document]
Comment (sunnights) [reply backend/tests/services/rag/test_direct_injection.py]: 已解决

File: backend/app/services/rag/retrieval_service.py [hunk 4]
```diff
@@ -714,6 +754,186 @@ async def _execute_runtime_query(
...
+            logger.warning(
+                "[RAG] get_original_documents: no knowledge base IDs provided"
+            )
+            return []
+
+        max_text_length = settings.MAX_EXTRACTED_TEXT_LENGTH
+        records: List[Dict[str, Any]] = []
+
+        # Case 1: Query by document IDs with KB scope filter
+        if document_ids:
```
Comment (kissghosts) [inline backend/app/services/rag/retrieval_service.py:799]: 这种不会多个 kb_ids 混合是吗？

File: backend/app/services/rag/retrieval_service.py [hunk 5]
```diff
@@ -714,6 +754,186 @@ async def _execute_runtime_query(
...
+            records = self._build_document_records(results, knowledge_base_ids)
+            logger.info(
+                "[RAG] get_original_documents completed: doc_ids=%s, document_count=%d",
+                document_ids,
+                len(records),
+            )
+            return records
+
+        # Case 2: Query each KB sequentially, stop on first truncation
+        for kb_id in knowledge_base_ids:
```
Comment (kissghosts) [inline backend/app/services/rag/retrieval_service.py:855]: 需要批量查出来吗？

File: backend/app/services/rag/retrieval_service.py (hunk 4 above)
Comment (sunnights) [reply backend/app/services/rag/retrieval_service.py:799]: 会混合查询，但返回结果中会标识每条记录的来源知识库。

查询层面： 确实是混合的，使用 IN 条件一次性查询多个知识库：
.filter(KnowledgeDocument.kind_id.in_(knowledge_base_ids))

返回层面： 每条记录都包含 knowledge_base_id 字段标识来源：
document_read_service.py:85
return {
    ...
    "kb_id": document.kind_id,  # 每条记录都有来源
}

retrieval_service.py:931
"knowledge_base_id": kb_id or result.get("kb_id"),  # 透传到最终结果

与主分支对比：
- 主分支：遍历每个 kb_id 逐个查询，显式设置 "knowledge_base_id": kb_id
- 当前分支：批量查询，从数据库记录中获取 kind_id

两种方式结果一致，调用方可以通过 knowledge_base_id 字段区分每条记录的来源知识库。

File: backend/app/services/rag/retrieval_service.py (hunk 5 above)
Comment (sunnights) [reply backend/app/services/rag/retrieval_service.py:855]: 当前主要沿用主分支的实现方式，没有做批量查询
knowledge_base_ids数量一般很少，可以之后再优化

**Review by qdaxb — APPROVED** [review]
